src/lib/gateway/GatewayStorage.ts

```diff
@@ -0,0 +1,137 @@
+import { Client } from '../types';
+import { Schema, SchemaJson } from '../schema/Schema';
+import { Provider } from '../structures/Provider';
+
+export class GatewayStorage {
+
+	/**
+	 * The client this gateway was created with.
+	 */
+	public readonly client: Client;
+
+	/**
+	 * The name of this gateway.
+	 */
+	public readonly name: string;
+
+	/**
+	 * The schema for this gateway.
+	 */
+	public readonly schema: Schema;
+
+	/**
+	 * Whether or not this gateway has been initialized.
+	 */
+	public ready = false;
+
+	/**
+	 * The provider's name that manages this gateway.
+	 */
+	private readonly _provider: string;
+
+	public constructor(client: Client, name: string, options: GatewayStorageOptions = {}) {
+		this.client = client;
+		this.name = name;
+		this.schema = options.schema || new Schema();
+		this._provider = options.provider || client.options.providers.default || '';
+	}
+
+	/**
+	 * The provider that manages this gateway's persistent data.
+	 */
+	public get provider(): Provider | null {
+		return this.client.providers.get(this._provider) || null;
```

Comment:
```suggestion
		return this.client.providers.get(this.provider) || null;
```

Reply:
Do you notice you just went into a Stack Overflow?

Reply:
Heck yeah, I sure love me some infinite recursion baby